src/app/ReadHandler.cpp

```diff
@@ -0,0 +1,198 @@
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines read handler for a CHIP Interaction Data model
+ *
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <app/MessageDef/EventPath.h>
+#include <app/ReadHandler.h>
+
+namespace chip {
+namespace app {
+CHIP_ERROR ReadHandler::Init(Messaging::ExchangeManager * apExchangeMgr, InteractionModelDelegate * apDelegate)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    // Error if already initialized.
+    VerifyOrExit(apExchangeMgr != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(apDelegate != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(mpExchangeMgr == nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(mpExchangeCtx == nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    mpExchangeMgr     = apExchangeMgr;
+    mpExchangeCtx     = nullptr;
+    mpDelegate        = apDelegate;
+    mSuppressResponse = true;
+    mGetToAllEvents   = true;
+    MoveToState(HandlerState::Initialized);
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+void ReadHandler::Shutdown()
+{
+    ClearExistingExchangeContext();
+    MoveToState(HandlerState::Uninitialized);
+    mpExchangeMgr = nullptr;
+    mpDelegate    = nullptr;
+}
+
+CHIP_ERROR ReadHandler::ClearExistingExchangeContext()
+{
+    if (mpExchangeCtx != nullptr)
+    {
+        mpExchangeCtx->Abort();
+        mpExchangeCtx = nullptr;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ReadHandler::OnReadRequest(Messaging::ExchangeContext * apExchangeContext, System::PacketBufferHandle aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferHandle response;
+
+    mpExchangeCtx = apExchangeContext;
+
+    err = ProcessReadRequest(std::move(aPayload));
+    SuccessOrExit(err);
+
+    MoveToState(HandlerState::Reportable);
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogFunctError(err);
+        Shutdown();
+    }
+
+    return err;
+}
+
+CHIP_ERROR ReadHandler::SendReportData(System::PacketBufferHandle aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReportData, std::move(aPayload),
+                                     Messaging::SendFlags(Messaging::SendMessageFlags::kNone));
+    SuccessOrExit(err);
+    MoveToState(HandlerState::ReportDataSending);
```

Comment:
Nothing ever takes us out of this state, violating the API contract about how we will shut ourselves down.  Again, seems like we could just close the exchange here, then shut ourselves down.

Reply:
we should shutdown after we receive status report from read client
currently we don't support status report, yes
we need shutdown immediately
i would add todo there
that is why i have one unused function in this PR
void ReadHandler::OnReportProcessingComplete()
{
    ChipLogDetail(DataManagement, "OnReportProcessingComplete");
    Shutdown();
}